### PR TITLE
Add storybook support

### PR DESCRIPTION
**Pull request by MrFlashAccount**
### Pull Request Description

This PR adds support for `storybook` and reimplements the GUI check workflows according to the modern practices.
Closes: https://github.com/enso-org/cloud-v2/issues/1474

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

### CODE REVIEW

File: .github/workflows/gui-checks.yml [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        with:
+          fetch-depth: 0
+
+      - name: 📦 Setup pnpm
+        uses: pnpm/action-setup@v4
+
+      - uses: actions/setup-node@v4
+        name: ⎔ Setup Node
+        with:
+          node-version: 20
```
Comment (Frizi) [inline .github/workflows/gui-checks.yml]: Node version is not correct.

File: .github/workflows/gui-checks.yml [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    runs-on:
+      - self-hosted
+      - Linux
+    timeout-minutes: 60
+    strategy:
+      fail-fast: false
+      max-parallel: 24
+      matrix:
+        shardIndex: [1, 2, 3, 4, 5, 6]
+        shardTotal: [6]
```
Comment (Frizi) [inline .github/workflows/gui-checks.yml:115]: Sharding will cause the same machine to run multiple test suites in parallel. Since we already had issues with timeouts caused by significant CPU load, I'm not sure if this is a good idea. This is especially bad if it causes the whole pnpm install and node/wasm-pack setup to be repeated.

File: .github/workflows/gui-checks.yml [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+          fetch-depth: 0
+          ref: ${{ github.event.pull_request.head.ref }}
+
+      - name: 📦 Setup pnpm
+        uses: pnpm/action-setup@v4
+
+      - uses: actions/setup-node@v4
+        name: ⎔ Setup Node
+        with:
+          node-version: 20
```
Comment (Frizi) [inline .github/workflows/gui-checks.yml]: Node version is not correct.

File: .github/workflows/gui-checks.yml [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+  checks: write
+
+jobs:
+  lint:
+    name: 👮 Lint GUI
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
```
Comment (Frizi) [inline .github/workflows/gui-checks.yml]: Why do we need to fetch all history here? Depth 1 (the default) is highly preferred.

File: .github/workflows/gui-checks.yml [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+      - name: 🧹 Lint
+        id: lint
+        continue-on-error: true
+        run: pnpm run ci:lint
+
+      - name: 🧪 Unit Tests
+        id: unit-tests
+        continue-on-error: true
+        run: pnpm run ci:test
```
Comment (Frizi) [inline .github/workflows/gui-checks.yml:73]: This might be fine, but just be aware that those jobs used to be running in parallel. Since most of those jobs are actually long-running single-threaded node processes, there wasn't much issue with doing that. With this new setup, they will all be serialized.

File: .github/workflows/gui-pull-request.yml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  changed-files:
+    runs-on: ubuntu-latest
+    name: 🔍 Detect changed files in GUI
+    outputs:
+      all_changed_files: ${{ steps.changed-files.outputs.all_changed_files }}
+      any_changed: ${{ steps.changed-files.outputs.any_changed }}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
```
Comment (Frizi) [inline .github/workflows/gui-pull-request.yml]: Again this fetch-depth setting is really bad here. You should limit the depth to `2`, that should be enough to detect changes.

File: app/common/src/appConfig.js [hunk 7]
```diff
@@ -46,24 +46,9 @@ export async function readEnvironmentFromFile() {
...
-    for (const key of expectedKeys) {
-      if (!(key in process.env)) {
-        missingKeys.push(key)
-      }
-    }
-    if (missingKeys.length !== 0) {
-      console.warn('Could not load `.env` file; disabling cloud backend.')
-      console.warn(`Missing keys: ${missingKeys.map(key => `'${key}'`).join(', ')}`)
-      console.error(error)
-    }
```
Comment (Frizi) [inline app/common/src/appConfig.js:66]: Why is this deleted?

File: pnpm-lock.yaml
Comment (Frizi) [inline pnpm-lock.yaml:1]: This amount of new dependencies is really bad for us for many reasons. We are constantly trying to reduce this file to absolute minimum, but this singular change adds almost 10% overhead that will be very tricky to get rid of. Many added dependencies are duplicates of things we already depend on, but with different versions. Please optimize the dependency tree so we don't have those duplicates where possible.

**Review by Frizi — CHANGES_REQUESTED** [review]
Let's split the PR into two individual parts:
- new actions workflows with file change detection
- storybook integration

The actions changes are a fine improvement over current setup, once some issues are fixed. They will just get deleted once we integrate bazel, since the whole thing will be redundant with it.

The storybook changes are not so clear cut, and we should discuss the tradeoffs and plans how we want to use it before integrating it. There are reasons why we very recently decided to **remove** existing storybook-like setup from the GUI project.

File: pnpm-lock.yaml
Comment (MrFlashAccount) [reply pnpm-lock.yaml:1]: This is an absolutely necessary, I added only essential deps and plugins for storybook

File: app/common/src/appConfig.js (hunk 7 above)
Comment (MrFlashAccount) [reply app/common/src/appConfig.js:66]: No longer valid, we now fails if env vars are not provided

File: .github/workflows/gui-checks.yml (hunk 2 above)
Comment (MrFlashAccount) [reply .github/workflows/gui-checks.yml:115]: We used to have pw setup where it spawns amount of workers equal the half of CPU cores and didn't have any queuing for the tasks.

I've reduces the amount of workers to 2 per shard and spawn 6 shards with total of 24 parallel shards per all PRs. I tested this approach yesterday and it seems like it improves the stability a lot

Comment (MrFlashAccount) [reply .github/workflows/gui-checks.yml:115]: Pnpm install is cached between jobs and installation takes 10secs. The rest of the time take the postinstall scripts (where we compile deps). Should be solved by bazel

File: .github/workflows/gui-checks.yml (hunk 5 above)
Comment (MrFlashAccount) [reply .github/workflows/gui-checks.yml:73]: Yeah, they run in serial now, but lint and unit testing depend on libs to be compiled atm, and lint/prettier have caches (which speed things up), also, one failed task won't cancel the whole workflow.

Making them fully parallel is not necessary because it's not the longest task out there

File: .github/workflows/gui-checks.yml (hunk 4 above)
Comment (MrFlashAccount) [reply .github/workflows/gui-checks.yml]: Yeah, here we can use the default fetch depth

**Comment by MrFlashAccount** [comment]
> The actions changes are a fine improvement over current setup, once some issues are fixed. They will just get deleted once we integrate bazel, since the whole thing will be redundant with it.

Disagree here. Bazel will only improve the existing setup, but not eliminate it completely. Sharding, linting, deploying will remain unchanged mostly.

**Comment by MrFlashAccount** [comment]
> The storybook changes are not so clear cut, and we should discuss the tradeoffs and plans how we want to use it before integrating it. There are reasons why we very recently decided to remove existing storybook-like setup from the GUI project.

We're not forcing you to use it, it's important for dasboard since we already have a lot of components/widgets with complex states and we need an isolated playground to test it. Screenshot tests and play functions are nice additions.

File: .github/workflows/gui-checks.yml (hunk 3 above)
Comment (MrFlashAccount) [reply .github/workflows/gui-checks.yml]: Did we solve the issue with sqlite3 installation on node 22?

File: .github/workflows/gui-pull-request.yml [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+      - name: Get changed files
+        id: changed-files
+        uses: tj-actions/changed-files@v44
+        with:
+          files: |
```
Comment (hubertp) [inline .github/workflows/gui-pull-request.yml:37]: what about `Cargo.*`, `pnpm*`, `corepack*`, `package.json` and so on?

**Review by hubertp — CHANGES_REQUESTED** [review]
I don't think the filter list is exhaustive

File: .github/workflows/gui-pull-request.yml (hunk 8 above)
Comment (MrFlashAccount) [reply .github/workflows/gui-pull-request.yml:37]: good catch. not sure about the cargo stuff though, but updates for lock file/lint configs and like should be listed

**Comment by MrFlashAccount** [comment]
@hubertp and @Frizi , I addressed your questions.

Still don't think that excluding CI from the PR is a good thing, if we want to revert stuff, we can simply delete the files.

File: .github/workflows/gui-checks.yml (hunk 3 above)
Comment (farmaazon) [reply .github/workflows/gui-checks.yml]: AFAIK yes, we now have no sqlite3 dependency.

File: pnpm-lock.yaml
Comment (farmaazon) [reply pnpm-lock.yaml:1]: So maybe we could look at other dependencies? Maybe bump them, so it will unify those having 3 versions?

**Comment by farmaazon** [comment]
> Still don't think that excluding CI from the PR is a good thing,

The main point is that there is no consensus about storybook (do we really need so many dependencies? Are there any alternatives?), and stuff improving CI is to be merged ASAP.

It's also quite simple - if reverting may be done by deleting files, moving to a new branch may be easily done by checkouting them.

**Comment by PabloBuchu** [comment]
@Frizi @farmaazon  can you elaborate over how this is redundant to the future bazel integration and more important since this PR introduces important changes for dashboard what is ETA for bazel?

Also I don't know the reasoning behind GUI not using the storybook but I know the reasoning for us to use it so its bit weird you are blocking us based on your assumptions

File: .github/workflows/gui-checks.yml (hunk 3 above)
Comment (MrFlashAccount) [reply .github/workflows/gui-checks.yml]: great, I did use the 20 version only because of that. replaced with reading from the .node-version file

**Comment by MrFlashAccount** [comment]
> The main point is that there is no consensus about storybook (do we really need so many dependencies? Are there any alternatives?), and stuff improving CI is to be merged ASAP.

Again, it's important for the dashboard, I'm not pushing you guys to use it (and we didn't setup chromatic for project-view)

Yes, alternatives are worse. And I don't think it's a huge thing in terms of the deps, since we use pnpm caches and restoration takes a couple of secs on CI. the rest time is taken by running postinstall scripts. I strongly suggest instead of focusing on amount of the deps (it's counterproductive), fix the postinstall stuff, see: https://github.com/enso-org/enso/actions/runs/11912481455/job/33197146572?pr=11522 (this task doesn't run postinstall)

**Review by hubertp — APPROVED** [review]
I still believe that filtering should be done in a separate PR for easier review/maintenance  but I won't be blocking PR because of that.

**Review by Frizi — APPROVED** [review]
This PR should be split, but given a big push to merge this, I'm fine with temporarily checking that in as is now, and possibly removing it later when it ends up causing trouble.
